aea/helpers/dialogue/base.py

```diff
@@ -182,70 +289,169 @@ def outgoing_extend(self, message: "Message") -> None:
 
     def incoming_extend(self, message: "Message") -> None:
         """
-        Extend the list of messages which keeps track of incoming messages.
+        Extend the list of incoming messages with 'message'
 
-        :param messages: a message to be added
+        :param message: a message to be added
         :return: None
         """
         self._incoming_messages.extend([message])
 
+    def outgoing_safe_extend(self, message: "Message") -> bool:
+        """
+        Extend the list of outgoing messages with 'message', if 'message' is valid
+
+        :param message: a message to be added
+        :return: True if message successfully added, false otherwise
+        """
+        if self.is_valid_next_message(message):
+            self.outgoing_extend(message)
+            return True
+        else:
+            return False
+
+    def incoming_safe_extend(self, message: "Message") -> bool:
+        """
+        Extend the list of incoming messages with 'message', if 'message' is valid
+
+        :param message: a message to be added
+        :return: True if message successfully added, false otherwise
+        """
+        if self.is_valid_next_message(message):
+            self.incoming_extend(message)
+            return True
+        else:
+            return False
+
+    def update(self, message: "Message") -> bool:
+        """
+        Extend the list of incoming/outgoing messages with 'message', if 'message' is valid
+
+        :param message: a message to be added
+        :return: True if message successfully added, false otherwise
+        """
+        if self.is_valid_next_message(message):
+            if message.is_incoming:
+                self.incoming_extend(message)
+            else:
+                self.outgoing_extend(message)
+            return True
+        else:
+            return False
```

Comment:
why not use your safe extend methods to avoid two levels of nesting?

Reply:
Done